File: go/cmd/aoc2022/day16.go
package main

import (
	"aoc/aoc"
	"container/list"
	"fmt"
	"regexp"
	"slices"
	"strconv"
	"strings"
	"time"
)

type day16 struct {
	valves     map[string]*Valve
	connection map[string][]string
}

func Day16() aoc.Solver {
	return &day16{
		valves:     make(map[string]*Valve),
		connection: make(map[string][]string),
	}
}

var RE_VALVE = regexp.MustCompile("Valve (\\S+) has flow rate=(\\d+); tunnels? leads? to valves? (.*)")

const PART1_MINUTES = 30
const PART2_MINUTES = 26
const START = "AA"

func (solver *day16) Parse(line string) {
	if line == "" {
		return
	}
	matches := RE_VALVE.FindStringSubmatch(line)
	if matches == nil {
		return
	}
	flow, ok := strconv.Atoi(matches[2])
	if ok != nil {
		panic("Error converting flow rate: " + line)
	}
	valve := Valve{
		name: matches[1],
		flow: flow,
		mask: 0,
	}
	solver.valves[valve.name] = &valve
	solver.connection[valve.name] = strings.Split(matches[3], ", ")
}

func (solver *day16) Solve() (*string, *string) {
	aoc.Info("Found %d valves to open in %d minutes", len(solver.valves), PART1_MINUTES)

	var valvesWithFlow = solver.calculateDistances()

	aoc.Info("Valves with flow: %d", len(valvesWithFlow))

	start, ok := solver.valves[START]
	if !ok {
		panic("Start valve not found in the input??")
	}
	var t0 = time.Now()
	var onePath = OnePathSolver{
		cache:          make(map[OnePathKey]OnePath),
		valvesWithFlow: valvesWithFlow,
	}
	bestPath1 := onePath.findPath(OnePath{visited: []*Valve{start}})
	var delta = time.Since(t0)
	var valvePath = solver.makePath(bestPath1.visited)
	aoc.Info("[1] Found max flow is %d: %s (%d cache hits, %d calls, %d cache size) [%.3fsec]",
		bestPath1.totalFlow, valvePath, onePath.cacheHits, onePath.calls, len(onePath.cache),
		delta.Seconds())

	t0 = time.Now()
	var twoPath = TwoPathSolver{
		cache:          make(map[TwoPathKey]TwoPath),
		valvesWithFlow: valvesWithFlow,
	}
	var bestPath2 = twoPath.findPath(InitTwoPath(len(valvesWithFlow)))
	delta = time.Since(t0)
	aoc.Info("[2] Found max flow is %d: %s / %s (%d cache hits, %d calls, %d cache size) [%.3fsec]",
		bestPath2.totalFlow, bestPath2.humanPath, bestPath2.elePath, twoPath.cacheHits, twoPath.calls,
		len(twoPath.cache), delta.Seconds())

	part1 := strconv.Itoa(bestPath1.totalFlow)
	part2 := strconv.Itoa(bestPath2.totalFlow)
	return &part1, &part2
}

func (solver *day16) calculateDistances() []*Valve {
	type distanceWrapper struct {
		cave     string
		distance int
	}
	var valvesWithFlow []*Valve
	start, ok := solver.valves[START]
	if !ok {
		panic("Start valve not found in the list??: ")
	}
	valvesWithFlow = append(valvesWithFlow, start)

	for _, valve := range solver.valves {
		if valve.flow > 0 {
			id := len(valvesWithFlow)
			valve.mask = 1 << (id - 1)
			valve.id = uint8(len(valvesWithFlow))
			valvesWithFlow = append(valvesWithFlow, valve)
		}
	}

	for _, curr := range valvesWithFlow {
		curr.tunnels = make([]uint8, len(valvesWithFlow))
		var visited = map[string]bool{curr.name: true}
		var queue = list.New()
		queue.PushBack(&distanceWrapper{curr.name, 0})
		for queue.Len() > 0 {
			var wrapper = queue.Remove(queue.Front()).(*distanceWrapper)
			for _, nextName := range solver.connection[wrapper.cave] {
				if _, ok := visited[nextName]; ok {
					continue
				}
				visited[nextName] = true
				next, ok := solver.valves[nextName]
				if !ok {
					panic(fmt.Sprintf("Valve connections for %s refer to an invalid value %s",
						wrapper.cave, nextName))
				}
				if next.flow > 0 || next.name == START {
					curr.tunnels[next.id] = uint8(wrapper.distance + 1)
				}
				queue.PushBack(&distanceWrapper{next.name, wrapper.distance + 1})
			}
		}
		aoc.Info("Distances for %s: %v", curr.name, curr.tunnels)
	}

	return valvesWithFlow
}

func (solver *day16) makePath(visited []*Valve) string {
	var path = "["
	for _, valve := range visited {
		path += valve.name + ","
	}
	return path[:len(path)-1] + "]"
}

type Valve struct {
	id      uint8
	tunnels []uint8
	name    string
	flow    int
	mask    uint
}

type OnePathSolver struct {
	cacheHits      int
	calls          int
	cache          map[OnePathKey]OnePath
	valvesWithFlow []*Valve
}

func (s *OnePathSolver) findPath(path OnePath) OnePath {
	s.calls += 1
	if (s.calls % 1000000) == 0 {
		aoc.Info("%d calls, %d cache hits...", s.calls, s.cacheHits)
	}
	var cacheKey = path.cacheKey()
	if cached, ok := s.cache[cacheKey]; ok {
		s.cacheHits += 1
		return path.merge(cached)
	}

	var cave = path.visited[len(path.visited)-1]
	var bestPath = path
	for _, valve := range s.valvesWithFlow {
		if valve.flow == 0 {
			continue
		}
		if (path.openValves & valve.mask) != 0 {
			continue
		}
		var distance = valve.tunnels[cave.id]
		var next = path.next(valve, distance)
		if next.elapsed >= PART1_MINUTES {
			continue
		}
		var subBest = s.findPath(next)
		if subBest.totalFlow > bestPath.totalFlow {
			bestPath = subBest
		}
	}

	s.cache[cacheKey] = bestPath.diff(path)
	return bestPath
}

type OnePathKey struct {
	name    string
	elapsed uint8
	valves  uint
}

type OnePath struct {
	visited    []*Valve
	openValves uint
	elapsed    uint8
	totalFlow  int
}

func (p *OnePath) cacheKey() OnePathKey {
	return OnePathKey{p.visited[len(p.visited)-1].name, p.elapsed, p.openValves}
}

func (p *OnePath) merge(other OnePath) OnePath {
	var visited = slices.Clone(p.visited)
	visited = append(visited, other.visited...)
	return OnePath{
		visited,
		p.openValves,
		p.elapsed + other.elapsed,
		p.totalFlow + other.totalFlow,
	}
}

func (p *OnePath) next(valve *Valve, distance uint8) OnePath {
	var visited = slices.Clone(p.visited)
	visited = append(visited, valve)
	var elapsed = p.elapsed + distance + 1
	var flow = (PART1_MINUTES - int(elapsed)) * valve.flow
	return OnePath{
		visited,
		p.openValves | valve.mask,
		elapsed,
		p.totalFlow + flow}
}

func (p *OnePath) diff(path OnePath) OnePath {
	var visited = p.visited[len(path.visited):len(p.visited)]
	return OnePath{
		visited,
		p.openValves,
		p.elapsed - path.elapsed,
		p.totalFlow - path.totalFlow,
	}
}

type TwoPathSolver struct {
	cacheHits      int
	calls          int
	cache          map[TwoPathKey]TwoPath
	valvesWithFlow []*Valve
}

func (s *TwoPathSolver) findPath(path TwoPath) TwoPath {
	s.calls += 1
	if (s.calls % 1000000) == 0 {
		aoc.Info("%d calls, %d cache hits...", s.calls, s.cacheHits)
	}

	var cacheKey = path.cacheKey()
	if cached, ok := s.cache[cacheKey]; ok {
		s.cacheHits += 1
		return path.merge(&cached)
	}

	var manPos = path.humanPath[path.humanPos]
	var manValve = s.valvesWithFlow[manPos]
	var elePos = path.elePath[path.elePos]
	var eleValve = s.valvesWithFlow[elePos]

	var bestPath = path
	for _, valve := range s.valvesWithFlow {
		if valve.flow == 0 {
			continue
		}
		// try to move both human and elephant towards the next valve
		if (path.openValves & valve.mask) != 0 {
			continue
		}
		// move human
		var distance = manValve.tunnels[valve.id]
		var next = path.nextHuman(valve, distance)
		if next.elapsed < PART2_MINUTES {
			var subBest = s.findPath(next)
			if subBest.totalFlow > bestPath.totalFlow {
				bestPath = subBest
			}
		}

		// move elephant
		distance = eleValve.tunnels[valve.id]
		next = path.nextElephant(valve, distance)
		if next.elapsed < PART2_MINUTES {
			var subBest = s.findPath(next)
			if subBest.totalFlow > bestPath.totalFlow {
				bestPath = subBest
			}
		}
	}

	s.cache[cacheKey] = bestPath.diff(&path)
	return bestPath
}

type TwoPathKey struct {
	humanPos     uint8
	humanElapsed uint8
	elePos       uint8
	eleElapsed   uint8
	valves       uint
}

type TwoPath struct {
	humanPath    []uint8
	humanPos     uint8
	humanElapsed uint8
	elePath      []uint8
	elePos       uint8
	eleElapsed   uint8
	openValves   uint
	elapsed      uint8
	totalFlow    int
}

func (p *TwoPath) cacheKey() TwoPathKey {
	return TwoPathKey{
		p.humanPath[p.humanPos],
		p.humanElapsed,
		p.elePath[p.elePos],
		p.eleElapsed,
		p.openValves,
	}
}

func (p *TwoPath) merge(other *TwoPath) TwoPath {
	var humanPath = slices.Clone(p.humanPath)
	copy(humanPath[p.humanPos+1:p.humanPos+other.humanPos+1],
		other.humanPath[0:other.humanPos+1])
	var elePath = slices.Clone(p.elePath)
	copy(elePath[p.elePos+1:p.elePos+other.elePos+1], other.elePath[0:other.elePos+1])
	return TwoPath{
		humanPath,
		p.humanPos + other.humanPos,
		p.humanElapsed + other.humanElapsed,
		elePath,
		p.elePos + other.elePos,
		p.eleElapsed + other.eleElapsed,
		p.openValves,
		p.elapsed + other.elapsed,
		p.totalFlow + other.totalFlow,
	}
}

func (p *TwoPath) nextHuman(valve *Valve, distance uint8) TwoPath {
	var humanPath = slices.Clone(p.humanPath)
	humanPath[p.humanPos+1] = valve.id
	var elapsed = p.humanElapsed + distance + 1
	var flow = (PART2_MINUTES - int(elapsed)) * valve.flow
	return TwoPath{
		humanPath,
		p.humanPos + 1,
		elapsed,
		slices.Clone(p.elePath),
		p.elePos,
		p.eleElapsed,
		p.openValves | valve.mask,
		max(elapsed, p.eleElapsed),
		p.totalFlow + flow,
	}

}

func (p *TwoPath) nextElephant(valve *Valve, distance uint8) TwoPath {
	var elePath = slices.Clone(p.elePath)
	elePath[p.elePos+1] = valve.id
	var elapsed = p.eleElapsed + distance + 1
	var flow = (PART2_MINUTES - int(elapsed)) * valve.flow
	return TwoPath{
		slices.Clone(p.humanPath),
		p.humanPos,
		p.humanElapsed,
		elePath,
		p.elePos + 1,
		elapsed,
		p.openValves | valve.mask,
		max(elapsed, p.humanElapsed),
		p.totalFlow + flow,
	}
}

func (p *TwoPath) diff(start *TwoPath) TwoPath {
	var humanPath = make([]uint8, len(p.humanPath))
	copy(humanPath, p.humanPath[start.humanPos+1:p.humanPos+1])
	var elePath = make([]uint8, len(p.elePath))
	copy(elePath, p.elePath[start.elePos+1:p.elePos+1])
	return TwoPath{
		humanPath,
		p.humanPos - start.humanPos,
		p.humanElapsed - start.humanElapsed,
		elePath,
		p.elePos - start.elePos,
		p.eleElapsed - start.eleElapsed,
		p.openValves,
		p.elapsed - start.elapsed,
		p.totalFlow - start.totalFlow,
	}
}

func InitTwoPath(size int) TwoPath {
	return TwoPath{
		humanPath: make([]uint8, size),
		elePath:   make([]uint8, size),
	}
}

File: java/src/main/java/aoc/year2022/Day16.java
package aoc.year2022;

import aoc.api.Results;
import aoc.api.Solver;
import aoc.util.Log;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Day16 implements Solver {

    private static final Pattern RE_VALVE = Pattern.compile("Valve (\\S+) has flow rate=(\\d+); tunnels? leads? to valves? (.*)");
    static final int PART1_MINUTES = 30;
    static final int PART2_MINUTES = 26;
    private static final String START = "AA";

    private final Map<String, Valve> valves = new HashMap<>();
    private final Map<String, List<String>> connections = new HashMap<>();

    @Override
    public void parse(String line) {
        if (line.isEmpty()) {
            return;
        }
        Matcher m = RE_VALVE.matcher(line);
        if (m.matches()) {
            var valve = new Valve(m);
            connections.put(valve.name, Arrays.stream(m.group(3).split(", ")).toList());
            valves.put(valve.name, valve);
        }
    }

    @Override
    public Results solve() {
        Log.info("Found %d valves to open in %d minutes", valves.size(), PART1_MINUTES);

        var valvesWithFlow = precacheValues();

        Log.info("Valves with flow: %d", valvesWithFlow.size());

        // part 1 - timed
        var t0 = System.currentTimeMillis();
        var onePathSolver = new OnePathSolver(valvesWithFlow);
        var bestPath1 = onePathSolver.findPath(new OnePath(valves.get(START)));
        var t1 = System.currentTimeMillis();
        var path = bestPath1.visited.stream().map(v -> v.name).toList();
        Log.info("[1] Found max flow is %d: %s (%d cache hits, %d calls, %d cache size) [%.3fsec]",
                bestPath1.totalFlow, path, onePathSolver.cacheHits, onePathSolver.calls,
                onePathSolver.cache.size(), (t1 - t0) / 1000.0);

        // part 2 - timed
        t0 = System.currentTimeMillis();
        var twoPathSolver = new TwoPathSolver(valvesWithFlow);
        var bestPath2 = twoPathSolver.findPath(new TwoPath(valvesWithFlow.size()));
        t1 = System.currentTimeMillis();
        var humanPath = valvePath(valvesWithFlow, bestPath2.humanPath,
                bestPath2.humanPathPos);
        var elePath = valvePath(valvesWithFlow, bestPath2.elePath, bestPath2.elePathPos);
        Log.info("[2] Found max flow is %d: %s / %s (%d cache hits, %d calls, %d cache size) [%.3fsec]",
                bestPath2.totalFlow, humanPath, elePath, twoPathSolver.cacheHits,
                twoPathSolver.calls, twoPathSolver.cache.size(), (t1 - t0) / 1000.0);
        return new Results(String.valueOf(bestPath1.totalFlow), String.valueOf(bestPath2.totalFlow));
    }

    private List<String> valvePath(List<Valve> valvesWithFlow, byte[] visited, byte length) {
        var path = new ArrayList<String>();
        for (int i = 0; i <= length; i++) {
            path.add(valvesWithFlow.get(visited[i]).name);
        }
        return path;
    }

    private List<Valve> precacheValues() {
        List<Valve> valvesWithFlow = new ArrayList<>();
        // add to the valves with flow and set the valve mask
        for (Valve valve : valves.values()) {
            if (valve.flow > 0) {
                valve.mask = 1 << valvesWithFlow.size();
                valvesWithFlow.add(valve);
                // id = 0 will be the START valve
                valve.id = (byte) valvesWithFlow.size();
            }
        }
        // always add the START valve as first
        valvesWithFlow.addFirst(valves.get(START));

        calculateDistances(valvesWithFlow);

        return valvesWithFlow;
    }

    private record DistanceWrapper(Valve cave, int distance) {
    }

    private void calculateDistances(List<Valve> valvesWithFlow) {
        // create a map associating valves and their connections;
        var tunnels = makeTunnelMap();

        for (Valve curr : valvesWithFlow) {
            // oh, java, when will you add a simple mutable map/list/set initializer?
            curr.tunnels = new byte[valvesWithFlow.size()];
            var visited = new HashSet<>(Set.of(curr.name));
            var queue = new ArrayList<DistanceWrapper>();
            queue.add(new DistanceWrapper(curr, 0));
            while (!queue.isEmpty()) {
                var wrapper = queue.removeFirst();
                var valve = wrapper.cave;
                for (Valve next : tunnels.get(valve.name)) {
                    if (visited.contains(next.name)) {
                        continue;
                    }
                    visited.add(next.name);
                    if (valvesWithFlow.contains(next)) {
                        curr.tunnels[next.id] = (byte) (wrapper.distance + 1);
                    }
                    queue.add(new DistanceWrapper(next, wrapper.distance + 1));
                }
            }
        }
    }

    private Map<String, List<Valve>> makeTunnelMap() {
        var tunnels = new HashMap<String, List<Valve>>();
        for (Entry<String, List<String>> conns : connections.entrySet()) {
            tunnels.put(conns.getKey(),
                    conns.getValue().stream().map(valves::get).toList());
        }
        return tunnels;
    }
}

class Valve {
    byte id = 0;
    byte[] tunnels;
    final String name;
    final int flow;
    int mask;

    public Valve(Matcher m) {
        name = m.group(1);
        flow = Integer.parseInt(m.group(2));
        mask = 0;
    }
}

class OnePathSolver {
    private final List<Valve> valvesWithFlow;
    int calls = 0;
    Map<OnePathKey, OnePath> cache = new HashMap<>();
    int cacheHits = 0;

    public OnePathSolver(List<Valve> valvesWithFlow) {
        this.valvesWithFlow = valvesWithFlow;
    }

    public OnePath findPath(OnePath path) {
        calls += 1;
        if ((calls % 1000000) == 0) {
            Log.info("%d calls, %d cache hits...", calls, cacheHits);
        }
        var cacheKey = path.cacheKey();
        if (cache.containsKey(cacheKey)) {
            cacheHits += 1;
            var cached = cache.get(cacheKey);
            return path.merge(cached);
        }

        var currValve = path.getPos();

        var bestPath = path;
        for (var valve : valvesWithFlow) {
            // ignore start - will not increase flow
            if (valve.flow == 0) {
                continue;
            }
            if ((path.openValves & valve.mask) != 0) {
                continue;
            }
            var distance = currValve.tunnels[valve.id];
            var next = path.next(valve, distance);
            if (next.elapsed >= Day16.PART1_MINUTES) {
                continue;
            }
            var subBest = findPath(next);
            if (subBest.totalFlow > bestPath.totalFlow) {
                bestPath = subBest;
            }
        }

        cache.put(cacheKey, bestPath.diff(path));
        return bestPath;
    }
}

record OnePathKey(String name, int elapsed, int valves) {
}

class OnePath {
    List<Valve> visited;
    int openValves;
    int elapsed;
    int totalFlow;

    public OnePath(Valve start) {
        visited = new ArrayList<>(List.of(start));
    }

    public OnePath(List<Valve> visited, int openValves, int elapsed, int totalFlow) {
        this.visited = visited;
        this.openValves = openValves;
        this.elapsed = elapsed;
        this.totalFlow = totalFlow;
    }

    public OnePathKey cacheKey() {
        return new OnePathKey(getPos().name, elapsed, openValves);
    }

    public OnePath merge(OnePath other) {
        var visited = new ArrayList<>(this.visited);
        visited.addAll(other.visited);
        return new OnePath(visited,
                openValves,
                elapsed + other.elapsed,
                totalFlow + other.totalFlow);
    }

    public OnePath next(Valve valve, byte distance) {
        var visited = new ArrayList<>(this.visited);
        visited.add(valve);
        var elapsed = this.elapsed + distance + 1;
        var flow = (Day16.PART1_MINUTES - elapsed) * valve.flow;
        return new OnePath(
                visited,
                openValves | valve.mask,
                elapsed,
                totalFlow + flow);
    }

    public OnePath diff(OnePath start) {
        var visited = new ArrayList<>(this.visited.subList(start.visited.size(), this.visited.size()));
        return new OnePath(
                visited,
                openValves,
                this.elapsed - start.elapsed,
                this.totalFlow - start.totalFlow);
    }

    public Valve getPos() {
        return visited.getLast();
    }
}

class TwoPathSolver {
    private final List<Valve> valvesWithFlow;
    int calls = 0;
    Map<TwoPathKey, TwoPath> cache;
    int cacheHits = 0;

    public TwoPathSolver(List<Valve> valvesWithFlow) {
        this.valvesWithFlow = valvesWithFlow;
        cache = HashMap.newHashMap(35_000_000);
    }

    public TwoPath findPath(TwoPath path) {
        calls += 1;
        if ((calls % 1000000) == 0) {
            Log.info("%d calls, %d cache hits...", calls, cacheHits);
        }

        var cacheKey = path.cacheKey();
        var cached = cache.get(cacheKey);
        if (cached != null) {
            cacheHits += 1;
            return path.merge(cached);
        }

        var manPos = path.getHumanPos();
        var manValve = valvesWithFlow.get(manPos);
        var elePos = path.getElephantPos();
        var eleValve = valvesWithFlow.get(elePos);

        var bestPath = path;
        for (var valve : valvesWithFlow) {
            // ignore start - will not increase flow
            if (valve.flow == 0) {
                continue;
            }
            // ignore valves already open
            if ((path.openValves & valve.mask) != 0) {
                continue;
            }
            // try to move both human and elephant towards the next valve
            // move human
            var distance = manValve.tunnels[valve.id];
            var next = path.nextHuman(valve, distance);
            if (next.elapsed < Day16.PART2_MINUTES) {
                var subBest = findPath(next);
                if (subBest.totalFlow > bestPath.totalFlow) {
                    bestPath = subBest;
                }
            }

            // move elephant
            distance = eleValve.tunnels[valve.id];
            next = path.nextElephant(valve, distance);
            if (next.elapsed < Day16.PART2_MINUTES) {
                var subBest = findPath(next);
                if (subBest.totalFlow > bestPath.totalFlow) {
                    bestPath = subBest;
                }
            }
        }

        cache.put(cacheKey, bestPath.diff(path));
        return bestPath;

    }
}

record TwoPathKey(byte humanPos, byte humanElapsed, byte elePos, byte eleElapsed, int valves) {
}

class TwoPath {
    byte[] humanPath;
    byte humanPathPos;
    byte humanElapsed;
    byte[] elePath;
    byte elePathPos;
    byte eleElapsed;
    int openValves;
    byte elapsed;
    int totalFlow;

    public TwoPath(int size) {
        humanPath = new byte[size + 1];
        humanPath[0] = 0;
        elePath = new byte[size + 1];
        elePath[0] = 0;
    }

    private TwoPath(byte[] humanPath, byte humanPathPos, byte humElapsed, byte[] elePath, byte elePathPos, byte eleElapsed, int openValves, byte elapsed, int totalFlow) {
        this.humanPath = humanPath;
        this.humanPathPos = humanPathPos;
        this.humanElapsed = humElapsed;
        this.elePath = elePath;
        this.elePathPos = elePathPos;
        this.eleElapsed = eleElapsed;
        this.openValves = openValves;
        this.elapsed = elapsed;
        this.totalFlow = totalFlow;
    }

    public TwoPathKey cacheKey() {
        return new TwoPathKey(
                getHumanPos(),
                humanElapsed,
                getElephantPos(),
                eleElapsed,
                openValves
        );
    }

    public TwoPath merge(TwoPath other) {
        var humPath = new byte[this.humanPath.length];
        System.arraycopy(this.humanPath, 0, humPath, 0, this.humanPathPos + 1);
        System.arraycopy(other.humanPath, 0, humPath, this.humanPathPos + 1, other.humanPathPos);
        var elePath = new byte[this.elePath.length];
        System.arraycopy(this.elePath, 0, elePath, 0, this.elePathPos + 1);
        System.arraycopy(other.elePath, 0, elePath, this.elePathPos + 1, other.elePathPos);
        return new TwoPath(
                humPath,
                (byte) (this.humanPathPos + other.humanPathPos),
                (byte) (this.humanElapsed + other.humanElapsed),
                elePath,
                (byte) (this.elePathPos + other.elePathPos),
                (byte) (this.eleElapsed + other.eleElapsed),
                this.openValves,
                (byte) (this.elapsed + other.elapsed),
                this.totalFlow + other.totalFlow
        );
    }

    public TwoPath nextHuman(Valve valve, int distance) {
        var humPath = new byte[this.humanPath.length];
        System.arraycopy(this.humanPath, 0, humPath, 0, this.humanPathPos + 1);
        humPath[this.humanPathPos + 1] = valve.id;
        var elePath = new byte[this.elePath.length];
        System.arraycopy(this.elePath, 0, elePath, 0, this.elePathPos + 1);
        var elapsed = this.humanElapsed + distance + 1;
        var flow = (Day16.PART2_MINUTES - elapsed) * valve.flow;
        return new TwoPath(
                humPath,
                (byte) (humanPathPos + 1),
                (byte) elapsed,
                elePath,
                elePathPos,
                this.eleElapsed,
                openValves | valve.mask,
                (byte) Math.max(elapsed, eleElapsed),
                totalFlow + flow
        );
    }

    public TwoPath nextElephant(Valve valve, int distance) {
        var humPath = new byte[this.humanPath.length];
        System.arraycopy(this.humanPath, 0, humPath, 0, this.humanPathPos + 1);
        var elePath = new byte[this.elePath.length];
        System.arraycopy(this.elePath, 0, elePath, 0, this.elePathPos + 1);
        elePath[this.elePathPos + 1] = valve.id;
        var elapsed = this.eleElapsed + distance + 1;
        var flow = (Day16.PART2_MINUTES - elapsed) * valve.flow;
        return new TwoPath(
                humPath,
                humanPathPos,
                humanElapsed,
                elePath,
                (byte) (elePathPos + 1),
                (byte) elapsed,
                openValves | valve.mask,
                (byte) Math.max(elapsed, humanElapsed),
                totalFlow + flow
        );
    }

    public TwoPath diff(TwoPath start) {
        var humanPath = new byte[this.humanPath.length];
        System.arraycopy(this.humanPath, start.humanPathPos + 1, humanPath, 0, this.humanPathPos - start.humanPathPos);
        var elePath = new byte[this.humanPath.length];
        System.arraycopy(this.elePath, start.elePathPos + 1, elePath, 0, this.elePathPos - start.elePathPos);
        return new TwoPath(
                humanPath,
                (byte) (this.humanPathPos - start.humanPathPos),
                (byte) (humanElapsed - start.humanElapsed),
                elePath,
                (byte) (this.elePathPos - start.elePathPos),
                (byte) (eleElapsed - start.eleElapsed),
                openValves,
                (byte) (elapsed - start.elapsed),
                totalFlow - start.totalFlow
        );
    }

    public byte getHumanPos() {
        return humanPath[humanPathPos];
    }

    public byte getElephantPos() {
        return elePath[elePathPos];
    }
}
